compiler/noirc_evaluator/src/acir/mod.rs

```diff
@@ -1278,30 +1278,13 @@ impl<'a> Context<'a> {
                 let slice_typ = dfg.type_of_value(slice_contents);
                 assert!(!slice_typ.is_nested_slice(), "ICE: Nested slice used in ACIR generation");
 
-                let acir_value = self.convert_value(slice_contents, dfg);
-                let (slice_length, result) = match acir_value {
-                    AcirValue::Var(_, _) => {
-                        unreachable!("ICE: cannot call `as_slice` on non-array type")
-                    }
-                    array @ AcirValue::Array(_) => {
-                        let array_len = if !slice_typ.contains_slice_element() {
-                            slice_typ.flattened_size() as usize
-                        } else {
-                            self.flattened_slice_size(slice_contents, dfg)
-                        };
-                        (array_len, array)
-                    }
-                    AcirValue::DynamicArray(source_array) => {
-                        let array =
-                            self.read_dynamic_array(source_array.block_id, source_array.len)?;
-                        let array = AcirValue::Array(array);
+                let slice_length = self.flattened_size(slice_contents, dfg);
+                let slice_length = self.acir_context.add_constant(slice_length);
 
-                        (source_array.len, array)
-                    }
-                };
+                let acir_value = self.convert_value(slice_contents, dfg);
+                let result = self.read_array(acir_value)?;
 
-                let slice_length = self.acir_context.add_constant(slice_length);
-                Ok(vec![AcirValue::Var(slice_length, AcirType::field()), result])
```

Comment:
Here we're just reading an array value and getting its length so I've replaced this with the `read_array` and `flattened_size` helper functions.